### ISSUE TITLE
Unrecognised type: reaction_added

### ISSUE THREAD

**Issue by peddie** [mpickering/slack-api#21]
{"type":"reaction_added","user":"<elided>","item":{"type":"message","channel":"<elided>","ts":"1438985103.000029"},"reaction":"no_good","event_ts":"1438985641.230240"}
    Unrecognised type: reaction_added
    Please report this failure to the github issue tracker

Another new constructor for `Event`?

**Comment by mpickering** [comment]
I just looked a bit more closely at this and I don't think that `Item` is the right type to parse `item` to.

**Comment by mpickering** [comment]
In the example JSON, the `type` is `message` but the docs state, 

> message: the item will have a message property containing a message object and a channel property > containing the channel ID for the message.

Did you remove the `message` field from the JSON or was it never there?

**Comment by mpickering** [comment]
Also I suspect that if `message` were there it would be a different type to what `Item` currently expects.

https://github.com/mpickering/slack-api/blob/master/src/Web/Slack/Types/Item.hs#L16

**Comment by peddie** [comment]
You're probably right about all this (I told you I had no clue!), but I am confused.  The example I see at https://api.slack.com/methods/reactions.list looks like the `item` field can contain various types, some of which correspond to the type definition and parsing I see in the code.  It looks to me like this field would be what the Haskell data type `Item` was written for.

What do you think it should be instead?

> Did you remove the message field from the JSON or was it never there?

I'm sorry, I don't think I changed anything regarding message fields, but I'm not 100% sure what you're asking.  Are you referring to the example I gave when I opened the ticket?  I didn't remove any fields from that.

**Comment by mpickering** [comment]
I got confused about what `Item` corresponded to. I think it should be `Item` but looking at the code for parsing `Item`s it seems that if the type is `message` then there should also be a field called `message` which is lacking from our example in the original ticket. 

Do you think that;s a bug in the SlackAPI? I don't know what a `reaction` is but looking at that JSON there's no way to tell which message the user reacted to? 

https://github.com/mpickering/slack-api/blob/master/src/Web/Slack/Types/Item.hs#L27

**Comment by peddie** [comment]
I see what you mean.  Yes, according to the docs, there should be a dictionary called `message`.  The only explanation I can see is that their API docs are inaccurate.  Let me generate a few more examples and see how consistent they are.

**Comment by peddie** [comment]
Indeed:

    {"type":"reaction_added","user":"<elided>","item":{"type":"message","channel":"<elided>","ts":"1439254477.000064"},"reaction":"+1","event_ts":"1439254518.050335"}
    key "message" not present
    Please report this failure to the github issue tracker

I'm not sure what to do about it.  I can't glean any additional context from the web interface, which makes it look like the reaction does correspond to one particular message.

**Comment by peddie** [comment]
Possibly related:

    {"type":"reaction_added","user":"U08KNTK3R","item":{"type":"file","file":"F08TBT85Q"},"reaction":"+1","event_ts":"1439257426.084541"}
    when expecting a File, encountered String instead
    Please report this failure to the github issue tracker

and

    {"type":"reaction_added","user":"U08KNTK3R","item":{"type":"file_comment","file_comment":"Fc08TC8MK3","file":"F08TBT85Q"},"reaction":"+1","event_ts":"1439257574.086176"}
    when expecting a File, encountered String instead
    Please report this failure to the github issue tracker

This suggests to me that the API docs are out of sync for other types of reaction, not just `message`.

**Comment by mpickering** [comment]
I don't know what to do here. I'm not surprised that this is broken, the API is broken in lots of subtle ways as you've already seen. Do you agree that the JSON for the message reaction looks incomplete? For the other two examples, I have seen something like this before. Maybe the best thing to do is to not use `Item` and introduce a new datatype called `ItemId` however unsatisfactory this is. 

Another way which would probably be better in the long run would be for the client to lookup these object ideas and replace them with the actual object they refer to but I don't have time to do this myself in the foreseeable future.

**Comment by peddie** [comment]
I agree that the JSON for the message reaction is incomplete.  I can introduce an `ItemId` (and possibly `FileId`?) newtype for now.  I also don't expect to have enough time to do the fancy version.

I will also try opening some more bugs against [their API docs](https://github.com/slackhq/slack-api-docs/issues); I see you've submitted quite a number already.

**Comment by peddie** [comment]
I'll also note that the JSON for the message reaction doesn't even say what message the reaction refers to, so we couldn't separately look up and inline the correct message if we tried!

**Comment by mpickering** [comment]
Right, that was my original concern. They replied to me on twitter and said that it should be a "simplified message object". Maybe the idea is to get the correct message from the timestamp?

**Comment by asayers** [comment]
The Slack docs seem to refer to these strange, incomplete-looking items as "embedded items". I had a crack at adding support for them in 9ae1a88; however, I'm not sure how complete my definitions are, since the slack docs don't really specify them. Some brief testing shows it seems to work.
